### PR TITLE
Make innerproduct similarity not supported

### LINKED ISSUE #380: [FEATURE] Lucene Based k-NN search support
**Is your feature request related to a problem?**
Recently Lucene started building in-house Approximate k-NN search in Java based on HNSW algorithm. k-NN plugin already has support for HNSW based out of C library namely NMSlib. By having the Lucene based HNSW support, we will be able to provide AKNN functionality independently outside of k-NN plugin with out having to rely on the NMSlib or FAISS. This will unlock the platform dependency issues because of C based libraries.

As a prerequisite we need to introduce new data type [dense_vector](https://github.com/opensearch-project/k-NN/issues/39) in OpenSearch core and build Field Mappers to index data from this field to Lucene graphs.

### PR DESCRIPTION
**Pull request by martin-gaievski**
Signed-off-by: Martin Gaievski <gaievski@amazon.com>

### Description
Remove Inner product (or Dot product) from list of supported similarity functions for Lucene engine. This is due to specifics of [Lucene implementation of Dot product](https://github.com/apache/lucene/blob/branch_9_2/lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java#L72-L74), score can be negative which violates OpenSearch concept. Tested by following test case. Plan is to test/address this issue properly and then open support in later versions.

for 2 vectors ```[-10.0, -10.0]``` and ```[10.0, 10.0]```  and search request ```[10, 10]``` here is the response

```
{
    "took": 1,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 2,
            "relation": "eq"
        },
        "max_score": 100.5,
        "hits": [
            {
                "_index": "my_index",
                "_id": "2",
                "_score": 100.5,
                "_source": {
                    "my_vector": [
                        10,
                        10
                    ]
                }
            },
            {
                "_index": "my_index",
                "_id": "1",
                "_score": -99.5,
                "_source": {
                    "my_vector": [
                        -10,
                        -10
                    ]
                }
            }
        ]
    }
}
``` 
 
### Issues Resolved
#380 
 
### Check List
- [X] All tests pass
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/knn/index/LuceneEngineIT.java [hunk 1]
```diff
@@ -68,8 +70,34 @@ public void testQuery_cosine() throws Exception {
         baseQueryTest(SpaceType.COSINESIMIL);
     }
 
-    public void testQuery_innerProduct() throws Exception {
-        baseQueryTest(SpaceType.INNER_PRODUCT);
+    public void testQuery_innerProduct_notSupported() throws Exception {
+        XContentBuilder builder = XContentFactory.jsonBuilder()
+                .startObject()
+                .startObject("properties")
```
Comment (naveentatikonda) [inline src/test/java/org/opensearch/knn/index/LuceneEngineIT.java:76]: Can we use constants

File: src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java [hunk 2]
```diff
@@ -349,6 +328,64 @@ public void testTypeParser_parse_fromKnnMethodContext_invalidDimension() throws
...
+        Settings settings = Settings.builder().put(settings(CURRENT).build()).build();
+
+        ModelDao modelDao = mock(ModelDao.class);
+        KNNVectorFieldMapper.TypeParser typeParser = new KNNVectorFieldMapper.TypeParser(() -> modelDao);
+
+        int efConstruction = 321;
+        int dimension = 133;
+        XContentBuilder xContentBuilderL1SpaceType = XContentFactory.jsonBuilder()
+                .startObject()
+                .field("type", "knn_vector")
```
Comment (naveentatikonda) [inline src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:344]: same here in this test

File: src/test/java/org/opensearch/knn/index/LuceneEngineIT.java (hunk 1 above)
Comment (martin-gaievski) [reply src/test/java/org/opensearch/knn/index/LuceneEngineIT.java:76]: ack

File: src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java (hunk 2 above)
Comment (martin-gaievski) [reply src/test/java/org/opensearch/knn/index/mapper/KNNVectorFieldMapperTests.java:344]: ack

**Review by jmazanec15 — APPROVED** [review]

**Review by naveentatikonda — APPROVED** [review]

**Pull request merged**